### PR TITLE
Add CLI equivalent to GUI's "export tracks"

### LINKED ISSUE #2254: CLI export tracks option
In the GUI, we have the options `Export` (which creates a single ogg/wav file) and `Export Tracks` (which creates an ogg/wav for each track).

On the CLI, we only seem to have the regular `Export` option (via the `--render` flag).
It would be nice to have a `--render-tracks` CLI option or something similar.

### PR DESCRIPTION
**Pull request by rcorre**
The GUI has an "export tracks" option that exports each track to a separate file.
This adds a `--render-tracks` CLI option to enable the same behavior from the command line.

In order to do this, I moved much of the rendering logic out of the gui code in `ExportProjectDialog` and into its own class called `RenderManager`.

Due to #588, `--render-tracks` will segfault just as `--render` does, but the output seems intact.

Resolves #2254.

### CODE REVIEW

**Comment by tresf** [comment]
Thanks.  FYI, https://github.com/LMMS/lmms/issues/2258

-Tres

**Comment by rcorre** [comment]
`--rendertracks`, then? Unless someone can come up with a nice, one-word description of what this does.

**Comment by tresf** [comment]
Render is assumed, right?  Perhaps just `--tracks`

**Comment by Wallacoloo** [comment]
@tresf the usage for the flag is `--render-tracks <project>` which makes it appear that you can use it without the `--render` flag, so I don't see how render can be assumed.

File: src/core/ProjectRenderer.cpp [hunk 1]
```diff
@@ -124,6 +124,15 @@ ProjectRenderer::ExportFileFormats ProjectRenderer::getFileFormatFromExtension(
 
 
 
+QString ProjectRenderer::getFileExtensionFromFormat(
+		ExportFileFormats _fmt )
```
Comment (Wallacoloo) [inline src/core/ProjectRenderer.cpp]: In all future code added to lmms, arguments should not be prefixed with `_` - as I understand, that is an old convention that is being phased out.

**Comment by tresf** [comment]
> [...] you can use it without the `--render` flag [...]

Then perhaps we change it.  What makes this different from `--loop` or any other modifier?

**Comment by rcorre** [comment]
In the GUI, loop and other modifiers are options within the export menu, while 'export tracks' is a totally different menu.

However that doesn't have to define the CLI interface, I'd be fine with
```
lmms --render --tracks proj.mmpz -f ogg -o output_dir
```
Thoughts?

**Comment by Wallacoloo** [comment]
@rcorre I believe your suggestion would require a more intelligent argument parser than what we have currently. The current system expects `--render <filename>` and would thus attempt to open the file `--tracks` if given a command like above.

If it could behave like `--render [options] <filename>`, that would be an improvement, since then options like `--tracks` or `--loop` are valid only when used within the `--render` command. It may be more clear to require the commands to be passed like `--render --tracks,--loop,proj.mmpz`, for example, such that everything after `--render` is still contained in one single `argv[i]` argument.

If anyone wants to do the work to support either the format you used above or the one where all render arguments are given as one `argv[i]`, go ahead. Otherwise, I'll leave the discussion to you and @tresf - I personally have no problem with the implementation you originally proposed in this PR, w/ the hyphen removed (`--rendertracks <filename>`).

**Comment by tresf** [comment]
No objections here either. :+1:

**Comment by Wallacoloo** [comment]
I checked out your branch, did a `rebase master` and then ran `./lmms --rendertracks ~/projects/example.mmpz -o ./example-tracks` after creating an `example-tracks` directory. It left the directory empty, and segfaulted.

The segfault was misleading - it looks like it's due to our rather poor separation between program core and gui - `AudioFileDevice` is attempting to show an error message about not being able to open the output file, but it tries to do so through the gui and crashes because the gui doesn't actually exist (one of the other open PRs sort-of resolves this).

It turned out the actual issue was just that the `-o` argument relies on the path having a trailing "/" when used with `--rendertracks`. How do you feel about making the trailing slash optional?

Also, `lmms --render <project>.mmpz` defaults to placing the output at `<project>.wav` when no `-o` flag is specified. Do you think something similar could be done for the `--rendertracks` option for consistency, rather than have it error out when no `-o` is given? Maybe place the tracks in `<project>` or `<project>.tracks`?

**Comment by Wallacoloo** [comment]
Oops, forgot to tag you @rcorre.

**Comment by rcorre** [comment]
@Wallacoloo, the trailing slash should definitely be optional, glad you caught that.
I'm not as sold on the default output directory. It worries me a bit that you could overwrite files without explicitly specifying a path (though, admittedly the chance of having a random directory named <project>.tracks that contains anything other than previously exported tracks is low).

On a related note, should -o create the output directory if it doesn't exist? It would be convenient and I can't think of a significant risk.

**Comment by Wallacoloo** [comment]
@rcorre fair points. Maybe just present a syntax error if no `-o` option is given then?

As far as implicitly creating the directory goes, I think it would be a good idea to create the deepest node if its parent directory exists. I'm not sure about creating the entire tree (i.e. `mkdir -p` style) though, as that might be undesirable if the user makes a typo in a long path. But it could certainly be argued either way - I'd be happy with either implementation.

**Comment by softrabbit** [comment]
> If it could behave like --render [options] <filename>, that would be an improvement, since then options like --tracks or --loop are valid only when used within the --render command. 

I think any order of options should be OK, as long as there is one non-option argument that refers to a project file. I.e. drop the parameter from the ```--render``` option and complain if there's no file to load. ```--render --output foo.wav project.mmpz --interpolation whatever```. Or you could at the end the render command put.

> It may be more clear to require the commands to be passed like --render --tracks,--loop,proj.mmpz, for example, such that everything after --render is still contained in one single argv[i] argument.

Hell no. *Maybe* after dropping the -- part, i.e.  ` `--render tracks,loop,proj.mmpz``` (and that's pretty ugly to me, too).

> Maybe just present a syntax error if no -o option is given then? As far as implicitly creating the directory goes, I think it would be a good idea to create the deepest node if its parent directory exists.

:+1: 

For the exact wording of options my preference is ```--render --tracks``` ahead of any ```--render-tracks``` variation. Preferably with the input file allowed at any position.

**Comment by rcorre** [comment]
@Wallacoloo: I haven't been able to repro the trailing slash issue. I'm using `QDir.filePath`, which I think should take care of adding a / if needed: https://github.com/rcorre/lmms/blob/master/src/core/RenderManager.cpp#L197

maybe something happened during the rebase? I'll try to get this branch back up to date

**Comment by Wallacoloo** [comment]
@rcorre Interesting - the problem seems to occur only when the output directory contains a literal ".". For example:
```
mkdir ./out.tracks
./lmms --rendertracks ~/lmms/projects/example.mmpz -o ./out.tracks
```
fails with a segfault.
However, 
```
mkdir ./out
./lmms --rendertracks ~/lmms/projects/example.mmpz -o ./out
```
succeeds.

The error could very well be in a part of the code not modified by you. My guess would be that the `-o` option tries to determine the render format by extracting (and removing) the argument's extension, but I haven't looked at the code so that's almost baseless.

**Comment by rcorre** [comment]
```
# bad
./lmms --rendertracks ~/lmms/projects/example.mmpz -o ./out.tracks
# good
./lmms --rendertracks ~/lmms/projects/example.mmpz -o ./out.tracks/
```
The plot thickens...

I guess this is the first lmms option to work with a directory argument from the CLI, so I'll have to make sure all the cases are handled -- all the right stuff is probably in QDir/QFile -- I just need to find the right way of applying it.

**Comment by rcorre** [comment]
Turns out we were calling `baseName` on the arg to `-o` as it was previously always a file.
If you passed a directory with no trailing '/' and a '.' it was misinterpreted as a file.

I've fixed this and rebased to master.

As a side note, the current behavior of `--render` always strips the extension:

    lmms --render file.mmpz -f ogg -o file.myext

will produce a file named `file.ogg`. I'm just preserving this behavior for now, but maybe it is a separate issue.

File: src/core/RenderManager.cpp [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+// Render the song into individual tracks
+void RenderManager::renderTracks()
+{
+	const TrackContainer::TrackList & tl = Engine::getSong()->tracks();
+
+	// find all currently unnmuted tracks -- we want to render these.
+	for( auto it = tl.begin(); it != tl.end(); ++it )
+	{
+		Track* tk = (*it);
```
Comment (Wallacoloo) [inline src/core/RenderManager.cpp]: In the future, feel free to use range-based iterators, i.e. `for (Track * tk : tl) { ... }`. Saves a bit of redundancy.

Comment (rcorre) [inline src/core/RenderManager.cpp]: Oh, cool. I didn't even know those existed. I'm not normally a c++ programmer :)

**Comment by Wallacoloo** [comment]
> the current behavior of --render always strips the extension [...] I'm just preserving this behavior for now, but maybe it is a separate issue.

Good call - that'd be best handled as a separate issue.

I tested this pretty thoroughly & everything is working as expected on my machine. The code looks fine, so we're good to merge :)

Nice work!

**Comment by tresf** [comment]
FYI - Next, time, we should squash these into a single commit before merging to simplify our commit  history.

**Comment by rcorre** [comment]
duly noted

**Pull request merged**